### PR TITLE
Feature/nfdiv 1128 notify applicant2 apply for co

### PR DESCRIPTION
**Pull request by benouaer**
### JIRA link (if applicable) ###
- https://tools.hmcts.net/jira/browse/NFDIV-1128
- https://tools.hmcts.net/jira/browse/NFDIV-1129
- https://tools.hmcts.net/jira/browse/NFDIV-1130

### Change description ###
- send notification to applicant2 upon lapsing of the holding period, that they can apply for conditional order
- send reminders to applicants 1 and 2 if 14 days passed and they haven't applied

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[X] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/divorce/common/event/SubmitConditionalOrder.java [hunk 1]
```diff
@@ -67,7 +67,7 @@ public void configure(ConfigBuilder<CaseData, State, UserRole> configBuilder) {
 
         log.info("Submit conditional order about to submit callback invoked for case id: {}", details.getId());
 
-        details.getData().getConditionalOrder().setDateSubmitted(LocalDateTime.now(clock));
+        details.getData().getConditionalOrder().setApplicant1SubmittedDate(LocalDateTime.now(clock));
```
Comment (tomxelliott) [inline src/main/java/uk/gov/hmcts/divorce/common/event/SubmitConditionalOrder.java:70]: Will this be compatible with solicitor side?

Comment (tomxelliott) [reply src/main/java/uk/gov/hmcts/divorce/common/event/SubmitConditionalOrder.java:70]: Seeing as this event was also being used by ExUI if I remember correctly

File: src/functionalTest/java/uk/gov/hmcts/divorce/systemupdate/event/SystemNotifyApplicantDisputeFormOverdueFT.java [hunk 2]
```diff
@@ -41,6 +42,8 @@
         "classpath:request/casedata/ccd-callback-casedata-system-notify-applicant-dispute-form-overdue.json";
     private static final String RESPONSE = "classpath:responses/response-system-notify-applicant-dispute-form-overdue.json";
 
+    @Value("${submit_aos.dispute_offset_days}")
+    private int disputeDueDateOffsetDays;
```
Comment (linusnorton) [inline src/functionalTest/java/uk/gov/hmcts/divorce/systemupdate/event/SystemNotifyApplicantDisputeFormOverdueFT.java:46]: why does this need to be configurable?

File: src/main/java/uk/gov/hmcts/divorce/systemupdate/event/SystemRemindApplicantsApplyForCOrder.java [hunk 3]
```diff
@@ -40,10 +43,19 @@ public void configure(final ConfigBuilder<CaseData, State, UserRole> configBuild
     public AboutToStartOrSubmitResponse<CaseData, State> aboutToSubmit(CaseDetails<CaseData, State> details,
                                                                        CaseDetails<CaseData, State> beforeDetails) {
         CaseData data = details.getData();
-        notification.sendToApplicant1(data, details.getId());
-        notification.sendToApplicant2(data, details.getId());
-        data.getApplication().setJointApplicantsNotifiedCanApplyForConditionalOrder(YES);
-
+        if (AwaitingConditionalOrder.equals(details.getState())) {
+            notification.sendToApplicant1(data, details.getId(), true);
```
Comment (linusnorton) [inline src/main/java/uk/gov/hmcts/divorce/systemupdate/event/SystemRemindApplicantsApplyForCOrder.java:47]: might need to add check if app1 is represented

File: src/main/java/uk/gov/hmcts/divorce/systemupdate/event/SystemRemindApplicantsApplyForCOrder.java [hunk 4]
```diff
@@ -40,10 +43,19 @@ public void configure(final ConfigBuilder<CaseData, State, UserRole> configBuild
...
+            if (!data.getApplicationType().isSole()) {
+                notification.sendToApplicant2(data, details.getId(), true);
```
Comment (linusnorton) [inline src/main/java/uk/gov/hmcts/divorce/systemupdate/event/SystemRemindApplicantsApplyForCOrder.java:49]: might need to add check if app2 is represented

File: src/main/java/uk/gov/hmcts/divorce/systemupdate/event/SystemProgressHeldCase.java [hunk 5]
```diff
@@ -19,19 +24,45 @@
...
+        final Long caseId = details.getId();
+
+        if (caseData.getApplicant1().isRepresented()) {
+            log.info("For case id {} applicant is represented by solicitor hence sending conditional order notification email", caseId);
+            conditionalOrderNotification.sendToSolicitor(caseData, caseId);
+        } else {
+            log.info("20wk holding period elapsed for Case({}), notifying Joint Applicants they can apply for conditional order", caseId);
+            conditionalOrderNotification.sendToApplicant1(caseData, caseId, false);
+            if (!caseData.getApplicationType().isSole()) {
+                conditionalOrderNotification.sendToApplicant2(caseData, caseId, false);
```
Comment (linusnorton) [inline src/main/java/uk/gov/hmcts/divorce/systemupdate/event/SystemProgressHeldCase.java:60]: you need to check if app2 is represented

File: src/main/java/uk/gov/hmcts/divorce/common/event/SubmitConditionalOrder.java [hunk 6]
```diff
@@ -72,7 +72,7 @@ public void configure(ConfigBuilder<CaseData, State, UserRole> configBuilder) {
         log.info("Submit conditional order about to submit callback invoked for case id: {}", details.getId());
 
         CaseData data = details.getData();
-        data.getConditionalOrder().setDateSubmitted(LocalDateTime.now(clock));
+        data.getConditionalOrder().setApplicant1SubmittedDate(LocalDateTime.now(clock));
```
Comment (linusnorton) [inline src/main/java/uk/gov/hmcts/divorce/common/event/SubmitConditionalOrder.java:75]: if we're going to assume app1 is submitting (which we should) then you need to clean up the permissions for the event: 

```
.grant(CREATE_READ_UPDATE, APPLICANT_1_SOLICITOR, CREATOR, CITIZEN, APPLICANT_2)
```

should be: 

```
.grant(CREATE_READ_UPDATE, APPLICANT_1_SOLICITOR, CREATOR)
```

(app1 or app1 solicitor)

**Review by tomxelliott — APPROVED** [review]

**Comment by tomxelliott** [comment]
Have approved rebase onto CO Submitted work for NFDIV-1053

File: src/functionalTest/java/uk/gov/hmcts/divorce/systemupdate/event/SystemNotifyApplicantDisputeFormOverdueFT.java (hunk 2 above)
Comment (benouaer) [reply src/functionalTest/java/uk/gov/hmcts/divorce/systemupdate/event/SystemNotifyApplicantDisputeFormOverdueFT.java:46]: because it is a magic number, I thought it might be worth making it configurable in case policy/business decide to change it

File: src/main/java/uk/gov/hmcts/divorce/divorcecase/tab/CaseTypeTab.java [hunk 7]
```diff
@@ -192,9 +192,9 @@ private void buildServiceApplicationTab(ConfigBuilder<CaseData, State, UserRole>
     private void buildConditionalOrderTab(ConfigBuilder<CaseData, State, UserRole> configBuilder) {
         configBuilder.tab("conditionalOrder", "Conditional Order")
             .forRoles(CASE_WORKER, LEGAL_ADVISOR, SOLICITOR, SUPER_USER)
-            .showCondition("coDateSubmitted=\"*\"")
+            .showCondition("coApplicant1SubmittedDate=\"*\"")
```
Comment (linusnorton) [inline src/main/java/uk/gov/hmcts/divorce/divorcecase/tab/CaseTypeTab.java:195]: theoretically this needs to be more complex now as with a joint it's the latter date of app1 or app2, but we'll have to tackle it later

File: src/main/java/uk/gov/hmcts/divorce/common/event/SubmitConditionalOrder.java (hunk 6 above)
Comment (linusnorton) [reply src/main/java/uk/gov/hmcts/divorce/common/event/SubmitConditionalOrder.java:75]: Also, we need to consider the situation where app2 has submitted first. So the states probably need to be:

```
.forStates(ConditionalOrderDrafted, ConditionalOrderPending)
```

And we need a check on the return state so that if it's a joint application AND originalState == ConditionalOrderDrafted THEN state = ConditionalOrderPending ELSE AwaitingLegalAdvisorReferral

File: src/main/java/uk/gov/hmcts/divorce/systemupdate/event/SystemRemindApplicantsApplyForCOrder.java [hunk 8]
```diff
@@ -6,32 +6,35 @@
...
     @Override
     public void configure(final ConfigBuilder<CaseData, State, UserRole> configBuilder) {
 
         configBuilder
-            .event(SYSTEM_NOTIFY_APPLICANTS_CONDITIONAL_ORDER)
-            .forState(AwaitingConditionalOrder)
-            .name("Joint Applicants Apply for CO")
-            .description("Notify Joint Applicants they can apply for a Conditional Order")
+            .event(SYSTEM_REMIND_APPLICANTS_CONDITIONAL_ORDER)
+            .forStates(AwaitingConditionalOrder, ConditionalOrderPending)
```
Comment (linusnorton) [inline src/main/java/uk/gov/hmcts/divorce/systemupdate/event/SystemRemindApplicantsApplyForCOrder.java:35]: I think ConditionalOrderDrafted too

File: src/main/java/uk/gov/hmcts/divorce/systemupdate/event/SystemRemindApplicantsApplyForCOrder.java [hunk 9]
```diff
@@ -40,10 +43,19 @@ public void configure(final ConfigBuilder<CaseData, State, UserRole> configBuild
...
-        data.getApplication().setJointApplicantsNotifiedCanApplyForConditionalOrder(YES);
-
+        if (AwaitingConditionalOrder.equals(details.getState())) {
+            notification.sendToApplicant1(data, details.getId(), true);
+            if (!data.getApplicationType().isSole()) {
+                notification.sendToApplicant2(data, details.getId(), true);
+            }
+        } else {
+            if (Objects.isNull(data.getConditionalOrder().getApplicant1SubmittedDate())) {
+                notification.sendToApplicant1(data, details.getId(), true);
```
Comment (linusnorton) [inline src/main/java/uk/gov/hmcts/divorce/systemupdate/event/SystemRemindApplicantsApplyForCOrder.java:53]: ditto

File: src/main/java/uk/gov/hmcts/divorce/systemupdate/event/SystemRemindApplicantsApplyForCOrder.java [hunk 10]
```diff
@@ -40,10 +43,19 @@ public void configure(final ConfigBuilder<CaseData, State, UserRole> configBuild
...
+            } else {
+                notification.sendToApplicant2(data, details.getId(), true);
```
Comment (linusnorton) [inline src/main/java/uk/gov/hmcts/divorce/systemupdate/event/SystemRemindApplicantsApplyForCOrder.java:55]: ditto

File: src/main/java/uk/gov/hmcts/divorce/systemupdate/schedule/SystemProgressHeldCasesTask.java [hunk 11]
```diff
@@ -70,66 +65,45 @@ public void run() {
...
-                caseId
-            );
+    private void submitEvent(CaseDetails caseDetails, User user, String serviceAuth) {
+        try {
+            final CaseData caseData = objectMapper.convertValue(caseDetails.getData(), CaseData.class);
+
+            LocalDate dateOfIssue = caseData.getApplication().getIssueDate();
+            log.info("issueDate from caseDataMap {}", dateOfIssue);
+
+            if (dateOfIssue == null) {
```
Comment (linusnorton) [inline src/main/java/uk/gov/hmcts/divorce/systemupdate/schedule/SystemProgressHeldCasesTask.java:88]: why is this needed?

File: src/main/java/uk/gov/hmcts/divorce/systemupdate/schedule/SystemProgressHeldCasesTask.java [hunk 12]
```diff
@@ -70,66 +65,45 @@ public void run() {
...
+                log.error("Ignoring case id {} with created on {} and modified on {}, as issue date is null",
+                    caseDetails.getId(),
+                    caseDetails.getCreatedDate(),
+                    caseDetails.getLastModified()
+                );
+            } else {
+                if (holdingPeriodService.isHoldingPeriodFinished(dateOfIssue)) {
```
Comment (linusnorton) [inline src/main/java/uk/gov/hmcts/divorce/systemupdate/schedule/SystemProgressHeldCasesTask.java:95]: why is this needed? shouldn't the ES query do this?

File: src/main/java/uk/gov/hmcts/divorce/systemupdate/schedule/conditionalorder/SystemRemindApplicantsApplyForCOrderTask.java [hunk 13]
```diff
@@ -50,36 +52,44 @@
...
         final User user = idamService.retrieveSystemUpdateUserDetails();
         final String serviceAuthorization = authTokenGenerator.generate();
         try {
             final BoolQueryBuilder query =
                 boolQuery()
-                    .must(matchQuery(STATE, AwaitingConditionalOrder))
-                    .filter(rangeQuery(DUE_DATE).lte(LocalDate.now()))
+                    .must(
+                        boolQuery()
+                            .should(matchQuery(STATE, AwaitingConditionalOrder))
```
Comment (linusnorton) [inline src/main/java/uk/gov/hmcts/divorce/systemupdate/schedule/conditionalorder/SystemRemindApplicantsApplyForCOrderTask.java:68]: Add Drafted too

File: src/main/java/uk/gov/hmcts/divorce/systemupdate/event/SystemRemindApplicantsApplyForCOrder.java [hunk 14]
```diff
@@ -40,10 +43,19 @@ public void configure(final ConfigBuilder<CaseData, State, UserRole> configBuild
     public AboutToStartOrSubmitResponse<CaseData, State> aboutToSubmit(CaseDetails<CaseData, State> details,
                                                                        CaseDetails<CaseData, State> beforeDetails) {
         CaseData data = details.getData();
-        notification.sendToApplicant1(data, details.getId());
-        notification.sendToApplicant2(data, details.getId());
-        data.getApplication().setJointApplicantsNotifiedCanApplyForConditionalOrder(YES);
-
+        if (AwaitingConditionalOrder.equals(details.getState())) {
```
Comment (linusnorton) [inline src/main/java/uk/gov/hmcts/divorce/systemupdate/event/SystemRemindApplicantsApplyForCOrder.java:46]: AwaitingConditionalOrder OR ConditionalOrderDrafted

File: src/main/java/uk/gov/hmcts/divorce/divorcecase/tab/CaseTypeTab.java (hunk 7 above)
Comment (benouaer) [reply src/main/java/uk/gov/hmcts/divorce/divorcecase/tab/CaseTypeTab.java:195]: do we show the tab if either of them are present (for joint) or both?

Comment (linusnorton) [reply src/main/java/uk/gov/hmcts/divorce/divorcecase/tab/CaseTypeTab.java:195]: if joint then show oldest of (app1 / app2 submission date)
if sole just show app1

Comment (linusnorton) [reply src/main/java/uk/gov/hmcts/divorce/divorcecase/tab/CaseTypeTab.java:195]: too complicated for now I think

File: src/main/java/uk/gov/hmcts/divorce/systemupdate/schedule/SystemProgressHeldCasesTask.java (hunk 11 above)
Comment (benouaer) [reply src/main/java/uk/gov/hmcts/divorce/systemupdate/schedule/SystemProgressHeldCasesTask.java:88]: this was there before, so didn't remove it

File: src/main/java/uk/gov/hmcts/divorce/systemupdate/schedule/SystemProgressHeldCasesTask.java (hunk 12 above)
Comment (benouaer) [reply src/main/java/uk/gov/hmcts/divorce/systemupdate/schedule/SystemProgressHeldCasesTask.java:95]: haven't changed this, but you're right. I'll remove it

File: src/functionalTest/java/uk/gov/hmcts/divorce/systemupdate/event/SystemRemindApplicantsApplyForCOrderFT.java [hunk 15]
```diff
@@ -58,17 +63,21 @@ public void shouldPassValidationAndSendEmailsToApplicantAndRespondent() throws I
     @EnabledIfEnvironmentVariable(named = "ELASTIC_SEARCH_ENABLED", matches = "true")
     public void shouldSearchForCasesReadyToApplyForConditionalOrder() {
         final BoolQueryBuilder query = boolQuery()
-            .must(matchQuery(STATE, AwaitingConditionalOrder))
-            .filter(rangeQuery(DUE_DATE).lte(LocalDate.now()))
+            .must(
+                boolQuery()
+                    .should(matchQuery(STATE, AwaitingConditionalOrder))
+                    .should(matchQuery(STATE, ConditionalOrderPending))
```
Comment (linusnorton) [inline src/functionalTest/java/uk/gov/hmcts/divorce/systemupdate/event/SystemRemindApplicantsApplyForCOrderFT.java:69]: sorry, did we need ConditionOrderDrafted here too?

File: src/main/java/uk/gov/hmcts/divorce/common/event/SubmitConditionalOrder.java [hunk 16]
```diff
@@ -72,7 +67,7 @@ public void configure(ConfigBuilder<CaseData, State, UserRole> configBuilder) {
...
```
Comment (linusnorton) [inline src/main/java/uk/gov/hmcts/divorce/common/event/SubmitConditionalOrder.java:71]:
```
var state = details.getData().isSoleApplication() 
  ? AwaitingLegalAdvisorReferral
  : beforeDetails.getState() == ConditionalOrderDrafted
    ? ConditionalOrderPending
    : AwaitingLegalAdvisorReferral

File: src/main/java/uk/gov/hmcts/divorce/common/event/SubmitConditionalOrder.java [hunk 17]
```diff
@@ -72,7 +67,7 @@ public void configure(ConfigBuilder<CaseData, State, UserRole> configBuilder) {
...
 
         if (!data.getApplicant1().isRepresented()) {
```
Comment (linusnorton) [inline src/main/java/uk/gov/hmcts/divorce/common/event/SubmitConditionalOrder.java:72]: if (state == AwaitingLegalAdvisorReferral && !data.getApplicant1().isRepresented())

File: src/main/java/uk/gov/hmcts/divorce/common/event/SubmitConditionalOrder.java (hunk 16 above)
Comment (linusnorton) [reply src/main/java/uk/gov/hmcts/divorce/common/event/SubmitConditionalOrder.java:71]: then on the return 

```
return AboutToStartOrSubmitResponse.<CaseData, State>builder()
            .data(details.getData())
            .state(state)
            .build();
```

**Review by linusnorton — APPROVED** [review]
